CC-20611: Warehouse picking feature integration guide. (#1609)
spryker/spryker-docs: _includes/pbc/all/install-features/202304.0/install-warehouse-picking-feature.md

|-----------------------------------------|------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Warehouse User Management               | {{site.version}} | [Install the Warehouse User Management feature](/docs/scos/dev/feature-integration-guides/{{site.version}}/install-the-warehouse-user-management-feature.html)                     |
| Order Management + Inventory Management | {{site.version}} | [Order Management and Inventory Management feature](/docs/scos/dev/feature-integration-guides/{{site.version}}/install-the-order-management-and-inventory-management-feature.html) |
| Shipment                                | {{site.version}} | [Install the Shipment feature](/docs/scos/dev/feature-integration-guides/{{site.version}}/shipment-feature-integration.html)                                                       |

> | Picker User Login                       | {{site.version}} |                                                                                                                                                                                    |

Reviewer: What is it? Is it a future stand-alone feature? I can't find it in the ReleaseGroup app.

Author: This feature right now in development.

| ApiPickingListRequestAttributes | class    | created | src/Generated/Shared/Transfer/ApiPickingListRequestAttributes |
| ApiPickingListAttributes        | class    | created | src/Generated/Shared/Transfer/ApiPickingListAttributes        |
| ApiPickingListItemAttributes    | class    | created | src/Generated/Shared/Transfer/ApiPickingListItemAttributes    |
| ApiPickingListRequestAttributes | class    | created | src/Generated/Shared/Transfer/ApiPickingListRequestAttributes |
| ApiPickingListRequestAttributes | class    | created | src/Generated/Shared/Transfer/ApiPickingListRequestAttributes |
| ApiPickingListRequestAttributes | class    | created | src/Generated/Shared/Transfer/ApiPickingListRequestAttributes |
| Item.uuid                       | property | created | src/Generated/Shared/Transfer/Item                            |
| User.uuid                       | property | created | src/Generated/Shared/Transfer/User                            |
| Stock.pickingListStrategy       | property | created | src/Generated/Shared/Transfer/Stock                           |

{% endinfo_block %}

### 3) Set up configuration

> 1. To make the `picking-lists` and `picking-list-items` resources protected, adjust the protected paths configuration:

Reviewer: ```suggestion
1. To make the `picking-lists` and `picking-list-items` resources protected, adjust the protected paths' configuration:
```

        </subprocesses>

        <transitions>
            <!-- We will find the following transition and substitute it with the provided below. -->
            <!--            <transition happy="true">-->
            <!--                <source>waiting</source>-->
            <!--                <target>exported</target>-->
            <!--                <event>check giftcard purchase</event>-->
            <!--            </transition>-->
            <!-- / -->

            <transition happy="true">
                <source>waiting</source>
                <target>picking list generation scheduled</target>
                <event>picking list generation schedule</event>
            </transition>

            <transition happy="true">
                <source>picking finished</source>
                <target>exported</target>
                <event>finish picking</event>
            </transition>

        </transitions>

    </process>

    <process name="DummyPicking"
             file="DummySubprocess/DummyPicking01.xml"/> <!-- Include the subprocess file to be loaded by the state machine. -->

</statemachine>
```

{% info_block warningBox "Verification" %}

> 

Reviewer: Add an introductory sentence to make clear what the user verifies—for example, "Make sure that the OMS transition diagram shows expected transitions:"

> 1. Go to your demo store administration panel and navigate to the OMS section.

Reviewer: If this step happens in the Back Office, then accept my suggestion:
```suggestion
1. In the Back Office of your demo store, navigate to **Administration&nbsp;<span aria-label="and then">></span> OMS**.
```

| PickingListItemsBackendResourcePlugin | Registers the `picking-list-items` resource. |               | Spryker\Glue\PickingListsBackendApi\Plugin\GlueBackendApiApplication |

**src/Pyz/Glue/GlueBackendApiApplication/GlueBackendApiApplicationDependencyProvider.php**

```php
<?php

namespace Pyz\Glue\GlueBackendApiApplication;

use Spryker\Glue\GlueBackendApiApplication\GlueBackendApiApplicationDependencyProvider as SprykerGlueBackendApiApplicationDependencyProvider;
use Spryker\Glue\PickingListsBackendApi\Plugin\GlueBackendApiApplication\PickingListsBackendResourcePlugin;
use Spryker\Glue\PickingListsBackendApi\Plugin\GlueBackendApiApplication\PickingListItemsBackendResourcePlugin;

class GlueBackendApiApplicationDependencyProvider extends SprykerGlueBackendApiApplicationDependencyProvider
{
    /**
     * @return array<\Spryker\Glue\GlueApplicationExtension\Dependency\Plugin\ResourceInterface>
     */
    protected function getResourcePlugins(): array
    {
        return [
            new PickingListsBackendResourcePlugin(),
            new PickingListItemsBackendResourcePlugin()
        ];
    }
}

```

{% info_block warningBox "Verification" %}

> As a prerequisite, pass the following steps:

Reviewer: If this is a procedure, then make replace bullets with steps (1., 2., 3., etc)

* Attach a Back Office user to any warehouse you have in the system. For this, use the [Install the Warehouse User Management feature](/docs/scos/dev/feature-integration-guides/{{site.version}}/install-the-warehouse-user-management-feature.html) guide.
* Place an order in the system so that the product is in the warehouse where you added the user in the previous step.
* Obtain the access token of the warehouse user.
* Use the warehouse user access token as the request header: `Authorization: Bearer {{your access token}}`.

Make sure that you can send the following requests:

> To get a collection of available picking lists for a warehouse user:

Reviewer: incomplete sentence, add the second part to make it complete—for example:
```suggestion
To get a collection of available picking lists for a warehouse user, send the request:
```

Adjust all the sentences below that has the same mistake.

* `GET https://glue-backend.mysprykershop.com/picking-lists`

To get a single pick list for a warehouse user:

* `GET https://glue-backend.mysprykershop.com/picking-lists/{% raw %}{{{% endraw %}picking-list-uuid{% raw %}}{{% endraw %}`

To start a pickup operation by a warehouse user:

* `PATCH https://glue-backend.mysprykershop.com/picking-lists/{% raw %}{{{% endraw %}picking-list-uuid{% raw %}}{{% endraw %}`

     ```json
     {
         "data": {
             "type": "picking-lists",
             "attributes": {
               "action": "startPicking"
             }
         }
     }
     ```

> To get a collection of the pick list items for a particular pick list:

Reviewer: It seems you use inconsistent terminology. Is it "pick list" or "picking list"? 
Decide which one is correct and make respective adjustments across all documents.

Author: The "picking list" is chosen.

            <!--            <transition happy="true">-->
            <!--                <source>waiting</source>-->
            <!--                <target>exported</target>-->
            <!--                <event>check giftcard purchase</event>-->
            <!--            </transition>-->
            <!-- / -->

            <transition happy="true">
                <source>waiting</source>
                <target>picking list generation scheduled</target>
                <event>picking list generation schedule</event>
            </transition>

            <transition happy="true">
                <source>picking finished</source>
                <target>exported</target>
                <event>finish picking</event>
            </transition>

        </transitions>

    </process>

    <process name="DummyPicking"
             file="DummySubprocess/DummyPicking01.xml"/> <!-- Include the subprocess file to be loaded by the state machine. -->

</statemachine>
```

{% info_block warningBox "Verification" %}

Make sure that the OMS transition diagram shows expected transitions:

> * In the Back Office of your demo store, navigate to **Administration&nbsp;<span aria-label="and then">></span> OMS**, you will find the process `DummyPayment01`, click on it to see the diagram.

Reviewer: ```suggestion
1. In the Back Office of your demo store, navigate to **Administration&nbsp;<span aria-label="and then">></span> OMS**, you will find the process `DummyPayment01`, click on it to see the diagram.
```

> * Make sure that the OMS transition diagram shows a possible transition from `waiting` to `picking list generation scheduled` and from `picking finished` to `exported`.

Reviewer: ```suggestion
2. Make sure that the OMS transition diagram shows a possible transition from `waiting` to `picking list generation scheduled` and from `picking finished` to `exported`.
```

{% endinfo_block %}

### 4) Import warehouse picking list strategies

1. Prepare your data according to your requirements using our demo data:

**data/import/common/common/warehouse.csv**

```csv
name,is_active,picking_list_strategy
Warehouse1,1,multi-shipment
Warehouse2,1,multi-shipment
Warehouse3,0,multi-shipment
```

> In case you are using the marketplace setup:

Reviewer: ```suggestion
If you are using the marketplace setup:
```

**data/import/common/common/marketplace/warehouse.csv**

```csv
name,is_active,picking_list_strategy
Warehouse1,1,multi-shipment
Warehouse2,1,multi-shipment
Warehouse3,0,multi-shipment
Spryker MER000001 Warehouse 1,1,multi-shipment
Video King MER000002 Warehouse 1,1,multi-shipment
Budget Cameras MER000005 Warehouse 1,1,multi-shipment
Sony Experts MER000006 Warehouse 1,1,multi-shipment
```

| Column                | REQUIRED? | Data Type | Data Example   | Data Explanation                                                     |
|-----------------------|-----------|-----------|----------------|----------------------------------------------------------------------|
| name                  | mandatory | string    | Warehouse1     | Name of the warehouse.                                               |
| is_active             | mandatory | bool      | 1              | Defines if the warehouse is active.                                  |
| picking_list_strategy | optional  | string    | multi-shipment | Defines the picking generation strategy used for the given warehouse |

2. Import data:

```bash
console data:import stock
```

{% info_block warningBox “Verification” %}

> Make sure that:

Reviewer: ```suggestion
Make sure the defined picking list strategies are applied correctly by finding them in the `spy_stock` database table.
```

> 

Reviewer: ```suggestion
```

> * The defined picking list strategies are applied correctly by finding them in the `spy_stock` database table.

Reviewer: ```suggestion
```

|-------------------------------------------------|----------------------------------------------------------|---------------|---------------------------------------------------------------------------------------------|
| MultiShipmentPickingListGeneratorStrategyPlugin | Generates the picking lists based on warehouse strategy. |               | Spryker\Zed\PickingListMultiShipmentPickingStrategyExample\Communication\Plugin\PickingList |

**\Pyz\Zed\PickingList\PickingListDependencyProvider.php**

```php
<?php

namespace Pyz\Zed\PickingList;

use Spryker\Zed\PickingListMultiShipmentPickingStrategyExample\Communication\Plugin\PickingList\MultiShipmentPickingListGeneratorStrategyPlugin;
use Spryker\Zed\PickingList\PickingListDependencyProvider as SprykerPickingListDependencyProvider;

class PickingListDependencyProvider extends SprykerPickingListDependencyProvider
{
    /**
     * @return list<\Spryker\Zed\PickingListExtension\Dependency\Plugin\PickingListGeneratorStrategyPluginInterface>
     */
    protected function getPickingListGeneratorStrategyPlugins(): array
    {
        return [
            new MultiShipmentPickingListGeneratorStrategyPlugin(),
        ];
    }
}
```

> 4. To enable the Backend API, register the plugins:

Reviewer: ```suggestion
3. To enable the Backend API, register the plugins:
```

class GlueBackendApiApplicationDependencyProvider extends SprykerGlueBackendApiApplicationDependencyProvider
{
    /**
     * @return array<\Spryker\Glue\GlueApplicationExtension\Dependency\Plugin\ResourceInterface>
     */
    protected function getResourcePlugins(): array
    {
        return [
            new PickingListsBackendResourcePlugin(),
            new PickingListItemsBackendResourcePlugin()
        ];
    }
}

```

{% info_block warningBox "Verification" %}

As a prerequisite, we have to pass the following steps:

1. Attach a back office user to any warehouse you have in the system (use the [Install the Warehouse User Management feature](/docs/scos/dev/feature-integration-guides/{{site.version}}/install-the-warehouse-user-management-feature.html) guide).
2. Place an order in the system so that the product is in the warehouse where you added the user in the previous step.
3. Obtain the access token of the warehouse user.
4. Use the warehouse user access token as the request header `Authorization: Bearer {{your access token}}`.

Make sure that you can send the following requests:

> To get a collection of available picking lists for a warehouse user, send the request:

Reviewer: ```suggestion
To get a collection of available picking lists for a warehouse user, send the request: `GET https://glue-backend.mysprykershop.com/picking-lists`.
```

> * `GET https://glue-backend.mysprykershop.com/picking-lists`

Reviewer: ```suggestion
```

> To get a single picking list for a warehouse user, send the request:

Reviewer: ```suggestion
To get a single picking list for a warehouse user, send the request: `GET https://glue-backend.mysprykershop.com/picking-lists/{% raw %}{{{% endraw %}picking-list-uuid{% raw %}}{{% endraw %}`.
```

> * `GET https://glue-backend.mysprykershop.com/picking-lists/{% raw %}{{{% endraw %}picking-list-uuid{% raw %}}{{% endraw %}`

Reviewer: ```suggestion
```

To start a pick up operation by a warehouse user, send the request:

> * `PATCH https://glue-backend.mysprykershop.com/picking-lists/{% raw %}{{{% endraw %}picking-list-uuid{% raw %}}{{% endraw %}`

Reviewer: ```suggestion
`PATCH https://glue-backend.mysprykershop.com/picking-lists/{% raw %}{{{% endraw %}picking-list-uuid{% raw %}}{{% endraw %}`
```

     ```json
     {
         "data": {
             "type": "picking-lists",
             "attributes": {
               "action": "startPicking"
             }
         }
     }
     ```

To get a collection of the picking list items for a particular picking list, send the request:

* `GET https://glue-backend.mysprykershop.com/picking-lists/{% raw %}{{{% endraw %}picking-list-uuid{% raw %}}{{% endraw %}/picking-list-items`

To get a single picking list item for a particular picking list, send the request:

* `GET https://glue-backend.mysprykershop.com/picking-lists/{% raw %}{{{% endraw %}picking-list-uuid{% raw %}}{{% endraw %}/picking-list-items/{% raw %}{{{% endraw %}picking-list-item-uuid{% raw %}{{{% endraw %}`

> To pick the picking list items, the endpoint works in a bulk mode, send the request:

Reviewer: ```suggestion
To pick the picking list items, send the following request. The endpoint works in a bulk mode:
```

{% endinfo_block %}

### 3) Set up configuration

1. To make the `picking-lists` and `picking-list-items` resources protected, adjust the protected paths configuration:

**src/Pyz/Shared/GlueBackendApiApplicationAuthorizationConnector/GlueBackendApiApplicationAuthorizationConnectorConfig.php**

```php
<?php

namespace Pyz\Shared\GlueBackendApiApplicationAuthorizationConnector;

use Spryker\Shared\GlueBackendApiApplicationAuthorizationConnector\GlueBackendApiApplicationAuthorizationConnectorConfig as SprykerGlueBackendApiApplicationAuthorizationConnectorConfig;

class GlueBackendApiApplicationAuthorizationConnectorConfig extends SprykerGlueBackendApiApplicationAuthorizationConnectorConfig
{
    /**
     * @return array<string, mixed>
     */
    public function getProtectedPaths(): array
    {
        return [
            '/\/picking-lists(?:\/[^\/]+)?\/?$/' => [
                'isRegularExpression' => true,
                'methods' => [
                    'get',
                    'getCollection',
                    'patch',
                ],
            ],

>             '/\/picking-lists/[^\/]+/picking-list-items(?:\/[^\/]+)?$' => [

Reviewer 2: I'm not clear a bit. Why does the description of these two roles not align with the main integration PR?
https://github.com/spryker/suite-nonsplit/pull/7525

Reviewer 3: Fixed
